frame/staking/src/lib.rs

```diff
@@ -530,57 +528,96 @@ impl<AccountId, Balance: HasCompact + Copy + Saturating + AtLeast32BitUnsigned +
 
 		(self, unlocking_balance)
 	}
-}
 
-impl<AccountId, Balance> StakingLedger<AccountId, Balance>
-where
-	Balance: AtLeast32BitUnsigned + Saturating + Copy,
-{
-	/// Slash the validator for a given amount of balance. This can grow the value
-	/// of the slash in the case that the validator has less than `minimum_balance`
-	/// active funds. Returns the amount of funds actually slashed.
+	/// Slash the staker for a given amount of balance. This can grow the value of the slash in the
+	/// case that either the active bonded or some unlocking chunks become dust after slashing.
+	/// Returns the amount of funds actually slashed.
 	///
-	/// Slashes from `active` funds first, and then `unlocking`, starting with the
-	/// chunks that are closest to unlocking.
-	fn slash(&mut self, mut value: Balance, minimum_balance: Balance) -> Balance {
-		let pre_total = self.total;
-		let total = &mut self.total;
-		let active = &mut self.active;
-
-		let slash_out_of =
-			|total_remaining: &mut Balance, target: &mut Balance, value: &mut Balance| {
-				let mut slash_from_target = (*value).min(*target);
-
-				if !slash_from_target.is_zero() {
-					*target -= slash_from_target;
-
-					// Don't leave a dust balance in the staking system.
-					if *target <= minimum_balance {
-						slash_from_target += *target;
-						*value += sp_std::mem::replace(target, Zero::zero());
-					}
-
-					*total_remaining = total_remaining.saturating_sub(slash_from_target);
-					*value -= slash_from_target;
-				}
-			};
-
-		slash_out_of(total, active, &mut value);
-
-		let i = self
-			.unlocking
-			.iter_mut()
-			.map(|chunk| {
-				slash_out_of(total, &mut chunk.value, &mut value);
-				chunk.value
-			})
-			.take_while(|value| value.is_zero()) // Take all fully-consumed chunks out.
-			.count();
+	/// # Note
+	///
+	/// This calls `Config::OnStakerSlash::on_slash` with information as to how the slash
+	/// was applied.
+	fn slash(
+		&mut self,
+		slash_amount: BalanceOf<T>,
+		minimum_balance: BalanceOf<T>,
+		slash_era: EraIndex,
+	) -> BalanceOf<T> {
+		use sp_staking::OnStakerSlash as _;
+
+		if slash_amount.is_zero() {
+			return Zero::zero()
+		}
 
-		// Kill all drained chunks.
-		let _ = self.unlocking.drain(..i);
+		let mut remaining_slash = slash_amount;
+		let pre_slash_total = self.total;
+
+		let era_after_slash = slash_era + 1;
+		let chunk_unlock_era_after_slash = era_after_slash + T::BondingDuration::get();
+
+		// Calculate the total balance of active funds and unlocking funds in the affected range.
+		let (affected_balance, slash_chunks_priority): (_, Box<dyn Iterator<Item = usize>>) = {
```

Comment:
Why is it important to have this priority here?  
Why dont you just start with the first chunk and slash from there?

Reply:
This is a good question. In the ideal scenario we want to slash the chunks created in the era after the slash up through the current era. This is because we want any delegator in the pool who was **actively bonded during the slash** to be slashed. 

For example:

Say we are at era 20 and a slash occurred at era 10. Delegators who unbonded in eras 0 -> 9 where definitely not actively bonded during the slash so we don't wan them to be affected. (Those who unbonded in era 10 likely did not know about the slash either, but we are not certain)